### PR TITLE
Sql pipelines

### PR DESCRIPTION
**Pull request by sjezewski**
Adds the following SQL support to putfile split:

- `Delimiter_SQL` type
- parsing support SQL PGDump data
- pokes SQL support through to `pachctl`

Also adds the header/footer abstraction to `put-file` split:

- so that `pachctl get-file /somedir` will return just the header + footer
- `pachctl get-file /somedir/000000003` will return the header + file contents + footer
- glob files like `pachctl get-file /somedir/*` will return the header + all child file contents + footer
- file sizes are reported to be the size of (header + content + footer)
- dir sizes are reported to be the size of (header + all child contents + footer)
- `put-header` and `put-footer` commands allow users to overwrite / delete the header/footer explicitly

This will come in handy for other formats (CSV,  other sql dump formats) as well. [Docs are in a separate PR](https://github.com/pachyderm/pachyderm/pull/3165)

### CODE REVIEW

File: src/server/pkg/sql/sql.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		done := false
+		for !done {
+			b, err := r.rd.ReadBytes('\n')
+			if err != nil {
+				if err == io.EOF {
+					return nil, fmt.Errorf("file does not contain row inserts")
+				}
+				return nil, err
+			}
+			if strings.HasPrefix(string(b), "COPY") {
```
Comment (jdoliner) [inline src/server/pkg/sql/sql.go]: This parsing feels a bit fragile, and writing non-fragile parsing yourself sounds quite annoying. Check this out and see if it's worth using: https://github.com/xwb1989/sqlparser

File: src/server/pkg/hashtree/interface.go [hunk 2]
```diff
@@ -100,6 +100,9 @@ type OpenHashTree interface {
 	// PutFile appends data to a file (and creates the file if it doesn't exist).
 	PutFile(path string, objects []*pfs.Object, size int64) error
 
+	// PutFileSplit appends data to a file (and creates the file if it doesn't exist).
+	PutFileSplit(path string, objects []*pfs.Object, size int64, header *pfs.Object, footer *pfs.Object, headerFooterSize int64) error
```
Comment (sjezewski) [inline src/server/pkg/hashtree/interface.go]: A bit on the fence about this signature / name. Adding 3 more args seems heavy - we could just create a struct for `metadata` that includes the header/footer/size ... and then call this method `PutFileWithMetadata`

File: src/server/pkg/sql/sql.go (hunk 1 above)
Comment (sjezewski) [reply src/server/pkg/sql/sql.go]: Discussed offline, but for posterity. I tried that lib as well as https://github.com/araddon/qlbridge but neither supported the SQL syntax we need. (Both barfed on `COPY` statements ... and qlbridge didn't like the pgdump format for `CREATE TABLE ...` either).

So for now, we're leaving our very dumb parser in place.

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -1917,6 +1919,60 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
...
+								ObjectHash: object.Hash,
+							}
+							return nil
+						})
+					}
+
+					if n > 0 {
+						// only increment here if the count is greater than 0,
+						// since we increment by default a few lines below
+						datumsWritten += n - int64(1)
```
[the remark this answers is not in the document]
Comment (jdoliner) [reply src/server/pfs/server/driver.go]: The only reason you have to do this is because your code doesn't fit in with the code around it. Each of the other cases reads a single instance of the delimited value off of the wire. Yours is reading multiple which is where the problems are coming from. Make it read a single value, and everything will make a lot more sense. Alternatively you could make the outside code expect to read multiple datums in each iteration of the loop... but I think that would be worse.

File: src/server/pachyderm_test.go [hunk 4]
```diff
@@ -7883,6 +7883,211 @@ func TestStatsDeleteAll(t *testing.T) {
...
+	expected = pgDumpHeader + rows[2] + pgDumpFooter
+	require.Equal(t, expected, buf.String())
+	require.Equal(t, len(expected), int(fileInfos[1].SizeBytes))
+	buf.Reset()
+	require.NoError(t, c.GetFile(commit.Repo.Name, commit.ID, "data", 0, 0, &buf))
+	require.Equal(t, pgDumpHeader+pgDumpFooter, buf.String())
+	dirFileInfo, err = c.InspectFile(commit.Repo.Name, commit.ID, "data")
+	require.NoError(t, err)
+	require.Equal(t, dirExpectedLen, int(dirFileInfo.SizeBytes))
+}
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: Would also be good to have tests that do multiple `PutFile` to the same SQL split to append stuff to it, including doing so with a different header and after files from the previous `PutFile` have been deleted.

Comment (jdoliner) [reply src/server/pachyderm_test.go]: It would also be good to have a test which creates multiple tables in the same repo `/table1` and `/table2` and then tries to get them out with `/**`, that should give you:

```
<table1 header>
<table1 row1>
<table1 row2>
...
<table1 footer>
<table2 header>
<table2 row1>
<table2 row2>
...
<table2 footer>
```

I think this would reveal some bugs in `GetFile`

File: src/server/pfs/server/driver.go [hunk 5]
```diff
@@ -2186,27 +2242,54 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
...
+		// e.g. this was a request for a single file e.g. /foo/bar.txt
+		dirNodePaths, err := tree.Glob(filepath.Dir(file.Path))
+		if err != nil {
+			return nil, err
+		}
+		if len(dirNodePaths) != 1 {
+			return nil, fmt.Errorf("invalid set of directory node paths - found %v expected 1 - this is probably a bug", len(dirNodePaths))
+		}
+		for key, node := range dirNodePaths {
+			paths[key] = node
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: This code seems overly specific to reading everything from a single table, I don't think it's going to work for reading from a single datum in a table or reading from multiple tables. We should hammer out some more test cases for this and then I think it'll be clearer what this code should look like.

File: src/server/pkg/hashtree/hashtree.go [hunk 6]
```diff
@@ -530,17 +534,24 @@ func (h *hashtree) putFile(path string, objects []*pfs.Object, overwriteIndex *p
...
 
 	// Add 'path' to parent (if it's new) & mark nodes as 'changed' back to root
+	updatedParent := false
 	return h.visit(path, func(node *NodeProto, parent, child string) error {
 		if node == nil {
 			node = &NodeProto{
 				Name:    base(parent),
 				DirNode: &DirectoryNodeProto{},
 			}
+			if !updatedParent {
```
Comment (jdoliner) [inline src/server/pkg/hashtree/hashtree.go]: This would be a bit clearer, and a bit shorter if you just check to see if `parent == filepath.Dir(path)` that's the intent here right? You only want to attach the header / footer to the direct parent of the path being put.

File: src/server/pfs/server/driver.go [hunk 7]
```diff
@@ -2518,7 +2603,14 @@ func (d *driver) applyWrite(key string, records *pfs.PutFileRecords, tree hashtr
...
 		for i, record := range records.Records {
-			if err := tree.PutFile(path.Join(key, fmt.Sprintf(splitSuffixFmt, i+int(indexOffset))), []*pfs.Object{{Hash: record.ObjectHash}}, record.SizeBytes); err != nil {
+			if err := tree.PutFileSplit(
+				path.Join(key, fmt.Sprintf(splitSuffixFmt, i+int(indexOffset))),
+				[]*pfs.Object{{Hash: record.ObjectHash}},
+				record.SizeBytes,
+				&pfs.Object{Hash: records.Header.ObjectHash},
+				&pfs.Object{Hash: records.Footer.ObjectHash},
+				records.Header.SizeBytes+records.Footer.SizeBytes,
+			); err != nil {
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: So is it accurate that we're passing the Header and Footer once for each datum we write out? That's probably not a big deal performance wise but it is a little inefficient and I think it might make more sense to instead have a special call, like maybe `tree.PutHeader` and `tree.PutFooter` so you can explicitly put the correct values into the directory once rather than doing it implicitly with each file you want to put.

File: src/server/pkg/sql/sql.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	return &PGDumpReader{
+		rd: r,
+	}
+}
+
+// ReadRows parses the pgdump file and populates the header and the footer
+// It returns EOF when done, and at that time both the Header and Footer will
+// be populated. Both header and footer are required. If either are missing, an
+// error is returned
+func (r *PGDumpReader) ReadRows(count int64) (rowsDump []byte, rowsRead int64, err error) {
```
Comment (jdoliner) [inline src/server/pkg/sql/sql.go]: I'm generally not a fan of using named return parameters, except for causes where you need reference them in closures (ie `retErr`) I've shot myself in the foot a bunch of times by inadvertently shadowing the parameters. I also don't think these are really making the code much shorter, they're just saving you writing a few `:=`s instead of `=`s I think. You also probably won't need `_err` if you don't use up that name in the return parameter.

File: src/server/pkg/sql/sql.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	if len(r.Header) == 0 {
+		err = r.readHeader()
+		if err != nil {
+			return nil, 0, err
+		}
+	}
```
Comment (jdoliner) [inline src/server/pkg/sql/sql.go]: Does this reader have support for just reading a series of statements w/o reading a header and footer? That seems like something we should be able to support as well.

File: src/server/pfs/server/driver.go (hunk 5 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Ah yea, very good point. We definitely need to handle those cases as well.

File: src/server/pkg/sql/sql.go (hunk 9 above)
Comment (sjezewski) [reply src/server/pkg/sql/sql.go]: What's the use case? For now I'm assuming that the input is a whole pgdump file that contains a single table's inserts. Right now we require both a header and footer, since that's what's needed to construct the table, etc. (Footer less so ... its just easier to require it for parsing reasons).

File: src/server/pkg/sql/sql.go (hunk 8 above)
Comment (sjezewski) [reply src/server/pkg/sql/sql.go]: Makes sense. Fixed.

File: src/server/pkg/hashtree/hashtree.go (hunk 6 above)
Comment (sjezewski) [reply src/server/pkg/hashtree/hashtree.go]: Makes sense. Fixed.

File: src/server/pfs/server/driver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Ah, somehow I missed that. Yes, this is much much simpler now.

**Comment by sjezewski** [comment]
Some example usage:

```
18-08-08[15:25:54]:pachyderm:0$pachctl put-file foo master tableA -f pgdump --split sql
warning: PFS destination "tableA" looks like a URL; did you mean -f tableA?
18-08-08[15:26:15]:pachyderm:0$pachctl list-file foo master
NAME   TYPE SIZE     
tableA dir  1.354KiB 
18-08-08[15:26:30]:pachyderm:0$pachctl list-file foo master /tableA
NAME                     TYPE SIZE     
/tableA/0000000000000000 file 1.21KiB  
/tableA/0000000000000001 file 1.212KiB 
/tableA/0000000000000002 file 1.213KiB 
18-08-08[15:26:34]:pachyderm:0$pachctl get-file foo master /tableA/0000000000000000
-- PostgreSQL database dump
--

-- Dumped from database version 10.4 (Debian 10.4-2.pgdg90+1)
-- Dumped by pg_dump version 10.4 (Debian 10.4-2.pgdg90+1)

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET client_min_messages = warning;
SET row_security = off;

SET default_tablespace = '';

SET default_with_oids = false;

--
-- Name: company; Type: TABLE; Schema: public; Owner: postgres
--

CREATE TABLE public.company (
    id integer NOT NULL,
    name text NOT NULL,
    age integer NOT NULL,
    address character(50),
    salary real
);


ALTER TABLE public.company OWNER TO postgres;

--
-- Data for Name: company; Type: TABLE DATA; Schema: public; Owner: postgres
--

COPY public.company (id, name, age, address, salary) FROM stdin;
1	alice	100	1234 acme st                                      	1000000
\.


--
-- Name: company company_pkey; Type: CONSTRAINT; Schema: public; Owner: postgres
--

ALTER TABLE ONLY public.company
    ADD CONSTRAINT company_pkey PRIMARY KEY (id);


--
-- PostgreSQL database dump complete
--
18-08-08[15:26:45]:pachyderm:0$pachctl get-file foo master /tableA
-- PostgreSQL database dump
--

-- Dumped from database version 10.4 (Debian 10.4-2.pgdg90+1)
-- Dumped by pg_dump version 10.4 (Debian 10.4-2.pgdg90+1)

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET client_min_messages = warning;
SET row_security = off;

SET default_tablespace = '';

SET default_with_oids = false;

--
-- Name: company; Type: TABLE; Schema: public; Owner: postgres
--

CREATE TABLE public.company (
    id integer NOT NULL,
    name text NOT NULL,
    age integer NOT NULL,
    address character(50),
    salary real
);


ALTER TABLE public.company OWNER TO postgres;

--
-- Data for Name: company; Type: TABLE DATA; Schema: public; Owner: postgres
--

COPY public.company (id, name, age, address, salary) FROM stdin;
\.


--
-- Name: company company_pkey; Type: CONSTRAINT; Schema: public; Owner: postgres
--

ALTER TABLE ONLY public.company
    ADD CONSTRAINT company_pkey PRIMARY KEY (id);


--
-- PostgreSQL database dump complete
--
```

And note how you'd delete/reset the header/footer values. In particular, since we validate the pgdump header/footer format ... you need to use the `line` delimiter to actually delete them.

```
18-08-08[15:26:57]:pachyderm:0$touch empty; echo '' | pachctl put-file foo master tableA --header empty
warning: PFS destination "tableA" looks like a URL; did you mean -f tableA?
cannot specify header without also setting --split flag
18-08-08[15:28:06]:pachyderm:1$touch empty; echo '' | pachctl put-file foo master tableA --header empty --split sql
warning: PFS destination "tableA" looks like a URL; did you mean -f tableA?
Reading from stdin.
invalid header - missing row inserts
18-08-08[15:28:12]:pachyderm:1$touch empty; echo '' | pachctl put-file foo master tableA --header empty --split line
warning: PFS destination "tableA" looks like a URL; did you mean -f tableA?
Reading from stdin.
18-08-08[15:28:26]:pachyderm:0$pachctl get-file foo master /tableA
\.


--
-- Name: company company_pkey; Type: CONSTRAINT; Schema: public; Owner: postgres
--

ALTER TABLE ONLY public.company
    ADD CONSTRAINT company_pkey PRIMARY KEY (id);


--
-- PostgreSQL database dump complete
--
18-08-08[15:28:33]:pachyderm:0$touch empty; echo '' | pachctl put-file foo master tableA --footer empty --split line
warning: PFS destination "tableA" looks like a URL; did you mean -f tableA?
Reading from stdin.
18-08-08[15:28:42]:pachyderm:0$pachctl get-file foo master /tableA
cannot read directory, no header or footer
18-08-08[15:28:44]:pachyderm:1$
```

File: src/server/pachyderm_test.go (hunk 4 above)
Comment (sjezewski) [reply src/server/pachyderm_test.go]: Indeed it did. This is now a test case (not sql, but the generic header/footer content case).

File: src/client/pfs.go [hunk 10]
```diff
@@ -716,20 +716,21 @@ func (c APIClient) Compact() error {
 // NOTE: PutFileWriter returns an io.WriteCloser you must call Close on it when
 // you are done writing.
 func (c APIClient) PutFileWriter(repoName string, commitID string, path string) (io.WriteCloser, error) {
-	return c.newPutFileWriteCloser(repoName, commitID, path, pfs.Delimiter_NONE, 0, 0, nil)
+	w, err := c.newPutFileWriteCloser(repoName, commitID, path, pfs.Delimiter_NONE, 0, 0, nil, nil, nil)
+	return w, err
```
Comment (jdoliner) [inline src/client/pfs.go]: Could we keep this as 1 line?

File: src/client/pfs.go [hunk 11]
```diff
@@ -1035,19 +1039,29 @@ type putFileWriteCloser struct {
...
+	}
+	request := &pfs.PutFileRequest{
+		File:             NewFile(repoName, commitID, path),
+		Delimiter:        delimiter,
+		TargetFileDatums: targetFileDatums,
+		TargetFileBytes:  targetFileBytes,
+		OverwriteIndex:   overwriteIndex,
+		Header:           headerValue,
+		Footer:           footerValue,
+	}
```
Comment (jdoliner) [inline src/client/pfs.go]: Could we keep this struct inlined in the return value? I find that a bit more readable.

File: src/server/pfs/cmds/cmds.go [hunk 12]
```diff
@@ -647,6 +650,13 @@ $ pachctl put-file repo branch -i file
 # NOTE this URL can reference local files, so it could cause you to put sensitive
 # files into your Pachyderm cluster.
 $ pachctl put-file repo branch -i http://host/path
+
+# Put a pgdump file
+$ pachctl put-file repo branch tablename -f tablename.pgdump --split sql
+
+# Overwrite header/footer for output generated by put-file split
+# To remove the header or footer, specify an empty file
+$ echo '' | pachctl put-file repo branch directory_containing_splitfiles --header=header.txt --footer=footer.txt -
```
Comment (jdoliner) [inline src/server/pfs/cmds/cmds.go]: This is a slightly awkward case that we may want to think about expanding a bit. Having to pipe in `echo ''` definitely isn't super pretty. You can achieve a similar effect by passing `-f /dev/null`, not sure if that's much better. Maybe we should consider adding `put-header` and `put-footer` which just update the header and footer, and those won't have any weird interactions with the existing `put-file` flags. They could just be different pachctl commands and still send `put-file` requests under the hood.

Comment (jdoliner) [reply src/server/pfs/cmds/cmds.go]: Also this example seems to be invalid because it doesn't have a `--split` flag.

File: src/server/pfs/server/driver.go [hunk 13]
```diff
@@ -1867,6 +1869,7 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
 		return err
 	}
 
+	limiter := limit.New(putObjectConcurrency)
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: It's not a particularly big deal but why did you move this up here? I generally like to keep declarations close to where they're used.

File: src/server/pachyderm_test.go [hunk 14]
```diff
@@ -7883,6 +7883,298 @@ func TestStatsDeleteAll(t *testing.T) {
...
+	require.NoError(t, c.GetFile(commit.Repo.Name, commit.ID, "/data/*", 0, 0, &buf))
+	// glob file is non deterministic ... so we need to check against all permutations of the rows:
+	validDumps := []string{
+		pgDumpHeader + strings.Join(rows, "") + pgDumpFooter,
+		pgDumpHeader + strings.Join([]string{rows[0], rows[2], rows[1]}, "") + pgDumpFooter,
+		pgDumpHeader + strings.Join([]string{rows[1], rows[0], rows[2]}, "") + pgDumpFooter,
+		pgDumpHeader + strings.Join([]string{rows[1], rows[2], rows[0]}, "") + pgDumpFooter,
+		pgDumpHeader + strings.Join([]string{rows[2], rows[0], rows[1]}, "") + pgDumpFooter,
+		pgDumpHeader + strings.Join([]string{rows[2], rows[1], rows[0]}, "") + pgDumpFooter,
+	}
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: Doesn't `GetFile` with a glob return things in sorted order? It certainly looks like there's some sorting going on in there.

File: src/server/pfs/server/driver.go [hunk 15]
```diff
@@ -2186,25 +2265,77 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
...
+			return nil, err
+		}
+		if len(dirNodePaths) != 1 {
+			return nil, fmt.Errorf("invalid set of directory node paths - found %v expected 1 - this is probably a bug", len(dirNodePaths))
+		}
+		for key, node := range dirNodePaths {
+			paths[key] = node
+			sortedPaths = append(sortedPaths, key)
+		}
+	}
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: I don't total understand what this block is doing. Why do we error if `len(dirNodePaths) != 1`? That doesn't seem like correct semantics, ie if someone did `get-file */0000000`, which gets the first row of each table at the top level, it seems like this would error if there were multiple directories, but that's a perfectly valid `get-file`.

At a high level I'm a bit confused about why you think you need separate code paths for the case where your glob pattern matches a directory and the case where it doesn't. It seems like you should be able to handle everything with the stack approach.

File: src/server/pfs/server/driver.go [hunk 16]
```diff
@@ -2186,25 +2265,77 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
...
+	sort.Strings(sortedPaths)
+	footers := stack.New()
+	directories := stack.New()
+	for _, path := range sortedPaths {
+		node := paths[path]
+		thisDir := directories.Peek()
+		if thisDir != nil && !strings.HasPrefix(path, thisDir.(string)) {
+			// We've proceeded past the current directory
+			footer := footers.Pop().(*pfs.Object)
+			if footer != nil {
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: Probably better to do this check before you push rather than after you pop, for deeply nested directory structures pushing and popping for every one could actually start to be a meaningful amount of memory moving around.

File: src/server/pkg/hashtree/hashtree.go [hunk 17]
```diff
@@ -497,43 +498,110 @@ func (h *hashtree) Finish() (HashTree, error) {
...
+		h.changed[path] = true
+		// Add 'path' to parent (if it's new) & mark nodes as 'changed' back to root
+		err := h.visit(path, func(node *NodeProto, parent, child string) error {
+			if node == nil {
+				node = &NodeProto{
+					Name:    base(parent),
+					DirNode: &DirectoryNodeProto{},
+				}
+				if parent == filepath.Dir(path) {
+					node.SubtreeSize = headerFooterSize
```
Comment (jdoliner) [inline src/server/pkg/hashtree/hashtree.go]: This seems incorrect, if `node.SubtreeSize != 0` then we'll be overwriting the existing value.

File: src/server/pkg/hashtree/hashtree.go [hunk 18]
```diff
@@ -497,43 +498,110 @@ func (h *hashtree) Finish() (HashTree, error) {
...
+					}
+				}
+				h.fs[parent] = node
+			}
+			insertStr(&node.DirNode.Children, child)
+			node.SubtreeSize += sizeDelta
+			h.changed[parent] = true
+			return nil
+		})
+		return err
```
Comment (jdoliner) [inline src/server/pkg/hashtree/hashtree.go]: Put the return on the line above, there's no reason to store an error in a variable if the next thing you're doing is returning it.

File: src/server/pkg/hashtree/hashtree.go [hunk 19]
```diff
@@ -497,43 +498,110 @@ func (h *hashtree) Finish() (HashTree, error) {
...
+func (h *hashtree) putFile(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64, header *pfs.Object, footer *pfs.Object, headerFooterSize int64) error {
 	path = clean(path)
 
 	// Detect any path conflicts before modifying 'h'
 	if err := h.visit(path, nop); err != nil {
 		return err
 	}
 
-	// Get/Create file node to which we'll append 'objects'
+	if len(objects) > 0 {
```
Comment (jdoliner) [inline src/server/pkg/hashtree/hashtree.go]: I think this code would be a lot cleaner and easier to understand if you didn't have 2 implementations, one for when there's objects and one for when there's not. I think it should be pretty easy to merge those into a single implementation.

**Comment by jdoliner** [comment]
Overall I think all of the major pieces are in place for this and it's coming together nicely. I think it would benefit from some more extensive testing, particular of some weirder cases with multiple tables in the same repo, deleting of some of the subfiles, tables within tables, etc. I'd also like to see some more examples of PGdumps from different versions and some corrupted ones to see how it responds to negative data. Also I think you'd benefit a lot from putting these tests in `src/server/pfs/server/server_test.go` it's much faster to run and you can easily drop into a debugger while you're running it. That's been the only way I've been able to get new pfs features with complicated semantics right, such as the original `--split` code.

File: src/server/pfs/cmds/cmds.go (hunk 12 above)
Comment (sjezewski) [reply src/server/pfs/cmds/cmds.go]: Sounds good. Made separate `put-header` and `put-footer` commands

File: src/server/pfs/server/driver.go (hunk 13 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: I put it there to raise its scope -- we use the limiter below when putting the objects for the header/footer ... around line 2012. I thought about omitting the use of a limiter there altogether since its just 2 additional puts per putfile ... but figured it was easy enough to use the limiter that was already there.

File: src/server/pachyderm_test.go (hunk 14 above)
Comment (sjezewski) [reply src/server/pachyderm_test.go]: Huh. Yea, it definitely does. I'd seen output w rows out of order during testing, but that must've been a bug in the way I was testing, because it's definitely ordered in the output. 

Anyway, this is much simpler now.

File: src/server/pfs/server/driver.go [hunk 20]
```diff
@@ -1917,6 +1920,40 @@ func (d *driver) putFile(ctx context.Context, file *pfs.File, delimiter pfs.Deli
...
+						if err != nil {
+							return err
+						}
+						records.Footer = &pfs.PutFileRecord{
+							SizeBytes:  size,
+							ObjectHash: object.Hash,
+						}
+						return nil
+					})
+				}
```
[the remark this answers is not in the document]
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Yup, easy change.

File: src/server/pfs/server/driver.go (hunk 13 above)
Comment (jdoliner) [reply src/server/pfs/server/driver.go]: Generally speaking raising a variables scope isn't a great thing to do unless you it's necessary. This is really all the same scope, just higher up in the scope. Still I think it's worth moving it down so it's declared close to where it's used.

File: src/server/pfs/server/driver.go (hunk 16 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: So, as-is, the code handles the case where we traverse a directory w a header and no footer. If I only add the footer if its non nil, then this check needs to change as well. We'd still need to check if its non nil to make sure we handle the case where the stack is empty (otherwise we'd panic when traversing a directory only w a header). But it would save some space on the stack. So I'll go ahead and add this test case and make these changes.

File: src/server/pfs/server/driver.go [hunk 21]
```diff
@@ -2223,25 +2274,78 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
...
+					continue
+				}
+				dirNode, err := tree.Get(dirPath)
+				if err != nil {
+					return nil, err
+				}
+				paths[dirPath] = dirNode
+			}
+		}
+	}
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: Looking forward to when we merge this with @brycemcanally's performance changes this is going to need to be implemented as a single pass over the paths, otherwise it's going to be painfully slow. I don't think it should be too hard to refactor this code to be a single pass. We can chat about how to do it in person.

File: src/server/pfs/server/driver.go (hunk 7 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Made this change

File: src/server/pfs/server/driver.go (hunk 21 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Ok, updated this so that we construct the `objects` array in a single pass, to support a streaming get file operation.

File: src/server/pfs/server/driver.go (hunk 13 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: It's not the same scope. If I leave the limiter declaration as-is it generates a compiler error. So you want me to declare two limiters? One for the put objects for the content and the other just for the header/footer?

File: src/server/pkg/sql/sql.go (hunk 9 above)
Comment (sjezewski) [reply src/server/pkg/sql/sql.go]: Per offline discussion, we decided this was out of scope. However, we don't have a good answer today on how we want to handle incremental SQL data ingestion. [Issue here](https://github.com/pachyderm/pachyderm/issues/3164)

File: src/server/pfs/server/driver.go (hunk 21 above)
Comment (jdoliner) [reply src/server/pfs/server/driver.go]: :+1:

**Comment by jdoliner** [comment]
All LGTM, I realized one more thing during reviewing which is that I think we need to update `CopyFile` to handle the `Header` and `Footer` we'll want to copy those over when someone does `CopyFile` on a directory with a `Header`/`Footer`. Sorry to move the goal posts again, this one should hopefully be pretty straighforward.
